internal/internal_event_handlers.go

```diff
@@ -977,6 +978,55 @@ func (weh *workflowExecutionEventHandlerImpl) handleChildWorkflowExecutionTermin
 	return nil
 }
 
+func (weh *workflowExecutionEventHandlerImpl) handleRequestCancelExternalWorkflowExecutionInitiated(event *m.HistoryEvent) error {
+	// For cancellation of child workflow only, we do not use cancellation ID
+	// for cancellation of all possible workflow, we have to use cancellation ID
+	attribute := event.RequestCancelExternalWorkflowExecutionInitiatedEventAttributes
+	workflowID := attribute.WorkflowExecution.GetWorkflowId()
+	cancellationID := string(attribute.Control)
+	weh.decisionsHelper.handleRequestCancelExternalWorkflowExecutionInitiated(event.GetEventId(), workflowID, cancellationID)
+	return nil
+}
+
+func (weh *workflowExecutionEventHandlerImpl) handleExternalWorkflowExecutionCancelRequested(event *m.HistoryEvent) error {
+	// For cancellation of child workflow only, we do not use cancellation ID
+	// for cancellation of all possible workflow, we have to use cancellation ID
+	attributes := event.ExternalWorkflowExecutionCancelRequestedEventAttributes
+	workflowID := attributes.WorkflowExecution.GetWorkflowId()
+	cancellationID := string(attributes.Control)
+	decision := weh.decisionsHelper.handleExternalWorkflowExecutionCancelRequested(attributes.GetInitiatedEventId(), workflowID)
+	if len(cancellationID) != 0 {
+		// for cancel external workflow, we need to set the future
+		cancellation := decision.getData().(*scheduledCancellation)
+		if cancellation.handled {
+			return nil
+		}
+		cancellation.handle(nil, nil)
+	}
+
+	return nil
+}
+
+func (weh *workflowExecutionEventHandlerImpl) handleRequestCancelExternalWorkflowExecutionFailed(event *m.HistoryEvent) error {
+	// For cancellation of child workflow only, we do not use cancellation ID
+	// for cancellation of all possible workflow, we have to use cancellation ID
+	attributes := event.RequestCancelExternalWorkflowExecutionFailedEventAttributes
+	workflowID := attributes.WorkflowExecution.GetWorkflowId()
+	cancellationID := string(attributes.Control)
+	decision := weh.decisionsHelper.handleRequestCancelExternalWorkflowExecutionFailed(attributes.GetInitiatedEventId(), workflowID)
+	if len(cancellationID) != 0 {
```

Comment:
it would be clearer if we could encapsulate this logic into one separate method. for example:
func isForChildWorkflow(event *m.HistoryEvent) bool, and change all the places where you do this check to that method call.

Reply:
sure